File: src/main/kotlin/sunsetsatellite/lang/sunlite/Type.kt
package sunsetsatellite.lang.sunlite

import sunsetsatellite.vm.sunlite.SLArray
import sunsetsatellite.vm.sunlite.SLBoundMethod
import sunsetsatellite.vm.sunlite.SLClass
import sunsetsatellite.vm.sunlite.SLClassInstance
import sunsetsatellite.vm.sunlite.SLClosure
import sunsetsatellite.vm.sunlite.SLFunction
import sunsetsatellite.vm.sunlite.SLNativeFunction
import sunsetsatellite.vm.sunlite.SLTable
import sunsetsatellite.vm.sunlite.SLType
import sunsetsatellite.vm.sunlite.SLUpvalue
import kotlin.math.sin

abstract class Type {

    open class Singular(val type: PrimitiveType, val ref: String = "") : Type() {

        override fun getName(): String {
            return type.name.lowercase()
        }

        override fun toString(): String {
            return getName()
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other != null) {
                if (other !is Singular) return false
            } else {
                return false
            }

            if(type == PrimitiveType.FLOAT || type == PrimitiveType.DOUBLE){
	            if(other.type != PrimitiveType.FLOAT && other.type != PrimitiveType.DOUBLE && numericOrder.containsKey(other.type)){
                    return true
                }
            }

            if(type == PrimitiveType.FLOAT && other.type == PrimitiveType.DOUBLE){
                return false
            } else if (type == PrimitiveType.DOUBLE && other.type == PrimitiveType.FLOAT){
                return true
            }

            if(numericOrder.containsKey(type) && numericOrder.containsKey(other.type)) {
                return numericOrder[other.type]!! <= numericOrder[type]!!
            }

            if (type != other.type) return false
            if (ref != other.ref && ref != "") return false

            return true
        }

        override fun hashCode(): Int {
            var result = type.hashCode()
            result = 31 * result + ref.hashCode()
            return result
        }
    }

    class Reference(type: PrimitiveType, ref: String, val returnType: Type, val params: List<Param> = listOf(), val typeParams: List<Param> = listOf()) : Singular(type, ref) {
        override fun getName(): String {
            return ref
        }

        override fun toString(): String {
            when (type) {
                PrimitiveType.FUNCTION -> {
                    return "${type.name.lowercase()}${if(typeParams.isNotEmpty()) "<${typeParams.joinToString()}>" else ""} '${ref}(${params.map { it.type }.joinToString()}): ${returnType}'"
                }
                PrimitiveType.ARRAY -> {
                    return "${type.name.lowercase()} '${returnType}'"
                }
                PrimitiveType.TABLE -> {
                    return "${type.name.lowercase()} '${typeParams[0].type} -> ${returnType}'"
                }
                PrimitiveType.OBJECT -> {
                    return "${type.name.lowercase()}${if(typeParams.isNotEmpty()) "<${typeParams.joinToString()}>" else ""}${if (ref != "") " '${ref}'" else ""}"
                }
                else -> {
                    return "${type.name.lowercase()}${if (ref != "") " '${ref}'" else ""}"
                }
            }
        }

        override fun equals(other: Any?): Boolean {
            if(other !is Reference) return false
            if(other.type != type) return false
            when (this.type) {
                PrimitiveType.FUNCTION -> {
                    if(this.returnType == other.returnType) {
                        if(this.params.size != other.params.size) return false
                        val types = params.map { it.type }
                        val otherTypes = other.params.map { it.type }
                        if(types.zip(otherTypes).any { !contains(it.second, it.first, currentInterpreter!!) }) return false
                        return true
                    }
                }
                PrimitiveType.OBJECT -> {
                    if(currentInterpreter == null) return false
                    if(ref == other.ref) return true
                    return traverseTypeHierarchy(other.ref)
                }
                PrimitiveType.ARRAY -> {
                    if(contains(returnType, other.returnType, currentInterpreter!!)) return true
                    return false
                }
                PrimitiveType.TABLE -> {
                    if(!contains(typeParams[0].type, other.typeParams[0].type, currentInterpreter!!)) return false
                    if(!contains(returnType, other.returnType, currentInterpreter!!)) return false
                    return true
                }
                PrimitiveType.CLASS -> {
                    if(this.ref == "") return true
                    if(this.ref == other.ref) return true
                }
                else -> {
                    return true
                }
            }
            return false
        }

        private fun traverseTypeHierarchy(other: String): Boolean {
            val parents = currentInterpreter?.collector?.typeHierarchy[other]
            val superclass = parents?.first
            val interfaces = parents?.second
            if(superclass == this.ref) return true
            if(interfaces != null && interfaces.contains(this.ref)) return true
            if((superclass == null || superclass == "<nil>" || superclass == "") && (interfaces == null || interfaces.isEmpty())) return false
            if(superclass != null){
                val success = traverseTypeHierarchy(superclass)
                if(success) return true
            }
            if(interfaces != null) {
                for (int in interfaces) {
                    val intSuccess = traverseTypeHierarchy(int)
                    if(intSuccess) return true
                }
            }
            return false
        }

        override fun hashCode(): Int {
            var result = type.hashCode()
            result = 31 * result + ref.hashCode()
            return result
        }

    }

    class Union(val types: List<Singular>) : Type() {
        override fun getName(): String {
            return types.joinToString(" | ")
        }

        override fun toString(): String {
            return getName()
        }
    }

    class Parameter(val name: Token) : Singular(PrimitiveType.GENERIC, name.lexeme) {
        override fun getName(): String {
            return name.lexeme
        }

        override fun toString(): String {
            return "type parameter '${name.lexeme}'"
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Parameter) return false

            if (name.lexeme != other.name.lexeme) return false

            return true
        }

        override fun hashCode(): Int {
            return name.hashCode()
        }


    }

    abstract fun getName(): String

    companion object {

        private val numericOrder: MutableMap<PrimitiveType, Int> = mutableMapOf(
            PrimitiveType.BYTE to 0, PrimitiveType.SHORT to 1, PrimitiveType.INT to 2, PrimitiveType.LONG to 3
        )

        var currentInterpreter: Sunlite? = null

        fun ofClass(name: String, params: List<Param> = listOf()): Reference {
            return Reference(PrimitiveType.CLASS, name, ofObject(name), params)
        }

        fun ofArray(elementType: Type): Reference {
            return Reference(PrimitiveType.ARRAY, "<array>", elementType, listOf())
        }

        fun ofTable(keyType: Type, valueType: Type): Reference {
            return Reference(PrimitiveType.TABLE, "<table>", valueType, listOf(), listOf(Param(Token.identifier("<key>"),keyType)))
        }

        fun ofFunction(name: String, returnType: Type, params: List<Param>): Reference {
            return Reference(PrimitiveType.FUNCTION, name, returnType, params)
        }

        fun ofObject(name: String): Reference {
            val reference = let {
                val reference = Reference(PrimitiveType.OBJECT, name, OBJECT, listOf())
                Reference(PrimitiveType.OBJECT, name, reference, listOf())
            }

            return reference
        }

        fun ofGenericObject(name: String, typeParams: List<Param>): Reference {
            val reference = let {
                val reference = Reference(PrimitiveType.OBJECT, name, OBJECT, listOf(), typeParams)
                Reference(PrimitiveType.OBJECT, name, reference, listOf(), typeParams)
            }

            return reference
        }

        fun ofGenericFunction(name: String, returnType: Type, params: List<Param>, typeParams: List<Param>): Reference {
            return Reference(PrimitiveType.FUNCTION, name, returnType, params, typeParams)
        }

        fun of(tokens: List<TypeToken>, sunlite: Sunlite, topmost: Boolean = true): Type {
            if(tokens.isEmpty()) {
                throw IllegalArgumentException("No types provided.")
            } else if(tokens.size > 1 && topmost) {
                throw IllegalArgumentException("There can be only one topmost type.")
            } else {
                if(topmost){
                    val topmostType = tokens.first()
                    if(topmostType.tokens.size == 1) {
                        val singleToken = topmostType.tokens.entries.first()
                        when (singleToken.key.type) {
                            TokenType.TYPE_FUNCTION -> {
                                val returnType = if(topmostType.typeParameters.isEmpty()) Type.NIL else of(listOf(topmostType.typeParameters.last()), sunlite,false)
                                val paramTokens = topmostType.typeParameters.dropLast(1)
                                val params = paramTokens.map { Param(Token.identifier(""), of(listOf(it), sunlite,false)) }
                                return ofFunction("", returnType, params)
                            }
                            TokenType.TYPE_ARRAY -> {
                                val elementType = if(topmostType.typeParameters.isEmpty()) NULLABLE_ANY else of(topmostType.typeParameters, sunlite,false)
                                return ofArray(elementType)
                            }
                            TokenType.TYPE_TABLE -> {
                                val keyType = if(topmostType.typeParameters.isEmpty()) NULLABLE_ANY else of(listOf(
                                    topmostType.typeParameters[0]
                                ), sunlite,false)
                                val valueType = if(topmostType.typeParameters.isEmpty()) NULLABLE_ANY else of(listOf(
                                    topmostType.typeParameters[1]
                                ), sunlite,false)
                                return ofTable(keyType, valueType)
                            }
                            TokenType.TYPE_CLASS -> {
                                if(topmostType.typeParameters.isEmpty()) {
                                    return ofClass("")
                                }
                                val className =
                                    topmostType.typeParameters.first().tokens.entries.first().key.lexeme
                                return ofClass(className)
                            }
                            TokenType.IDENTIFIER -> {
                                if(topmostType.typeParameters.isNotEmpty()) {
                                    val types = topmostType.typeParameters.map { of(listOf(it), sunlite,false) }
                                    val name = singleToken.key.lexeme
                                    val typeParams = sunlite.collector?.typeHierarchy[name]?.third
                                    return ofGenericObject(singleToken.key.lexeme, types.mapIndexed { i, it -> Param(Token.identifier(typeParams?.get(i) ?: "?"), it) })
                                }
                                return ofObject(singleToken.key.lexeme)
                            }
                            TokenType.TYPE_GENERIC -> {
                                return Parameter(topmostType.typeParameters.first().tokens.keys.first())
                            }
                            else -> {
                                return Singular(PrimitiveType.get(singleToken.key))
                            }
                        }
                    } else {
                        val types = topmostType.tokens.values.map { of(it, sunlite,false) as Singular }
                        return Union(types)
                    }
                } else {
                    if(tokens.size == 1) {
                        val singleType = tokens.first()
                        if(singleType.tokens.size == 1) {
                            val singleToken = singleType.tokens.entries.first()
                            when (singleToken.key.type) {
                                TokenType.TYPE_FUNCTION -> {
                                    val returnType = if(singleType.typeParameters.isEmpty()) NIL else of(listOf(singleType.typeParameters.last()),sunlite, false)
                                    val paramTokens = singleType.typeParameters.dropLast(1)
                                    val params = paramTokens.map { Param(Token.identifier(""), of(listOf(it), sunlite,false)) }
                                    return ofFunction("", returnType, params)
                                }
                                TokenType.TYPE_ARRAY -> {
                                    val elementType = if(singleType.typeParameters.isEmpty()) NULLABLE_ANY else of(singleType.typeParameters, sunlite,false)
                                    return ofArray(elementType)
                                }
                                TokenType.TYPE_TABLE -> {
                                    val keyType = if(singleType.typeParameters.isEmpty()) NULLABLE_ANY else of(listOf(
                                        singleType.typeParameters[0]
                                    ), sunlite,false)
                                    val valueType = if(singleType.typeParameters.isEmpty()) NULLABLE_ANY else of(listOf(
                                        singleType.typeParameters[1]
                                    ), sunlite,false)
                                    return ofTable(keyType, valueType)
                                }
                                TokenType.TYPE_CLASS -> {
                                    if(singleType.typeParameters.isEmpty()) {
                                        return ofClass("")
                                    }
                                    val className =
                                        singleType.typeParameters.first().tokens.entries.first().key.lexeme
                                    return ofClass(className)
                                }
                                TokenType.IDENTIFIER -> {
                                    if(singleType.typeParameters.isNotEmpty()) {
                                        val types = singleType.typeParameters.map { of(listOf(it), sunlite,false) }
                                        val name = singleToken.key.lexeme
                                        val typeParams = sunlite.collector?.typeHierarchy[name]?.third
                                        return ofGenericObject(singleToken.key.lexeme, types.mapIndexed { i, it -> Param(Token.identifier(typeParams?.get(i) ?: "?"), it) })
                                    }
                                    return ofObject(singleToken.key.lexeme)
                                }
                                TokenType.TYPE_GENERIC -> {
                                    return Parameter(singleType.typeParameters.first().tokens.keys.first())
                                }
                                else -> {
                                    return Singular(PrimitiveType.get(singleToken.key))
                                }
                            }
                        } else {
                            val types = singleType.tokens.map { of(it.value,sunlite, false) as Singular }
                            return Union(types)
                        }
                    } else {
                        val types = tokens.map { of(listOf(it), sunlite,false) as Singular }
                        return Union(types)
                    }
                }
            }
        }

        fun contains(type: Type, inType: Type, sunlite: Sunlite): Boolean {
            currentInterpreter = sunlite
            if(type == UNKNOWN) return false //can't statically determine the type
            if(inType is Union){
                if(type is Union) {
                    return inType.types.containsAll(type.types) || inType.types.contains(ANY)
                } else {
                    return inType.types.any { it.equals(type) } || inType.types.contains(ANY)
                }
            } else {
                if(type is Union) {
                    return type.types.contains(ANY)
                } else {
                    return inType.equals(type) || inType == ANY
                }
            }
        }


	    //TODO: better runtime type get for arrays and tables
        fun fromValue(value: Any?, sunlite: Sunlite): Type {
            return when (value) {
                is Type -> value
                is Param -> value.type
                is String -> STRING
                is Byte -> BYTE
                is Short -> SHORT
                is Int -> INT
                is Long -> LONG
                is Float -> FLOAT
                is Double -> DOUBLE //NUMBER
                is Boolean -> BOOLEAN
                is Unit -> NIL
                is SLClosure -> {
                    val function = value.function
                    ofFunction(function.name, function.returnType, function.params)
                }
                is SLBoundMethod -> {
                    val function = value.method.function
                    ofFunction(function.name, function.returnType, function.params)
                }
                is SLUpvalue -> {
                    fromValue(value.closedValue, sunlite)
                }
                is SLNativeFunction ->  ofFunction(value.name, value.returnType, listOf())
                is SLFunction -> ofFunction(value.name, value.returnType, value.params)
                is SLClass -> ofClass(value.name)
                is SLClassInstance -> ofObject(value.clazz.name)
                is SLArray -> ofArray(NULLABLE_ANY)
	            is SLTable -> ofTable(NULLABLE_ANY, NULLABLE_ANY)
                is SLType -> value.value
                else -> UNKNOWN
            }
        }

        val UNKNOWN = Singular(PrimitiveType.UNKNOWN)
        val NIL = Singular(PrimitiveType.NIL)
        val ANY = Singular(PrimitiveType.ANY)
        val CLASS = Singular(PrimitiveType.CLASS)
        val FUNCTION = Singular(PrimitiveType.FUNCTION)
        val OBJECT = Singular(PrimitiveType.OBJECT)
        val ARRAY = Singular(PrimitiveType.ARRAY)
        val TABLE = Singular(PrimitiveType.TABLE)
        val NULLABLE_ANY = Union(listOf(ANY, NIL))
        val BYTE = Singular(PrimitiveType.BYTE)
        val SHORT = Singular(PrimitiveType.SHORT)
        val INT = Singular(PrimitiveType.INT)
        val LONG = Singular(PrimitiveType.LONG)
        val FLOAT = Singular(PrimitiveType.FLOAT)
        val DOUBLE = Singular(PrimitiveType.DOUBLE)
        //val NUMBER = Singular(PrimitiveType.NUMBER)
        val STRING = Singular(PrimitiveType.STRING)
        val BOOLEAN = Singular(PrimitiveType.BOOLEAN)
    }
}

File: src/main/csharp/SunliteSharp/Core/Type.cs
using System.Runtime.InteropServices.ComTypes;
using System.Text;
using SunliteSharp.Core.Enum;
using SunliteSharp.Runtime;

namespace SunliteSharp.Core;

public abstract record Type
{
    public abstract string Name();

    public record Singular(PrimitiveType Type, string Ref = "") : Type
    {
        public override string Name()
        {
            return Type.ToString();
        }

        public override string ToString()
        {
            return Name();
        }
    }

    public record Reference(PrimitiveType Type, string Ref, Type ReturnType, List<Param> Params, List<Param> TypeParams) : Singular(Type, Ref)
    {
        public override string Name()
        {
            return Ref;
        }

        public override string ToString()
        {
            var s = new StringBuilder($"{Type.ToString()}");
            return Type switch
            {
                PrimitiveType.Function => s.Append($" '{Ref}({string.Join(", ",Params.Select(p => p.Type.Name()))}): {ReturnType}'").ToString(),
                PrimitiveType.Table => s.Append($" '{TypeParams[0].Type}' -> {ReturnType}").ToString(),
                PrimitiveType.Array => s.Append($" '{ReturnType}'").ToString(),
                _ => s.Append($" {(Ref != "" ? $"'{Ref}'" : "")}").ToString()
            };
        }

        public virtual bool Equals(Reference? other)
        {
            if (other is null) return false;
            if(other.Type != Type) return false;
            switch (Type)
            {
                case PrimitiveType.Function:
                    if (ReturnType == other.ReturnType)
                    {
                        if (Params.Count != other.Params.Count) return false;

                        IEnumerable<Type> types = Params.Select(p => p.Type);
                        IEnumerable<Type> otherTypes = other.Params.Select(p => p.Type);

                        return types.Zip(otherTypes, (t1, t2) => new { t1, t2 }).All(pair => Contains(pair.t2, pair.t1, sl));
                    }
                    break;

                case PrimitiveType.Object:
                    if (sl == null) return false;
                    return Ref == other.Ref || TraverseTypeHierarchy(other.Ref);

                case PrimitiveType.Array:
                    return Contains(ReturnType, other.ReturnType, sl);

                case PrimitiveType.Table:
                    return Contains(TypeParams[0].Type, other.TypeParams[0].Type, sl) && Contains(ReturnType, other.ReturnType, sl);

                case PrimitiveType.Class:
                    if (string.IsNullOrEmpty(Ref)) return true;
                    if (Ref == other.Ref) return true;
                    break;

                default:
                    return true;
            }

            return false;
        }
        
        private bool TraverseTypeHierarchy(string other)
        {
            (string, List<string>, List<string>)? parents = sl?.Collector?.TypeHierarchy.GetValueOrDefault(other);
            var superclass = parents?.Item1;
            List<string>? interfaces = parents?.Item2;

            if (superclass == Ref) return true;
            if (interfaces != null && interfaces.Contains(Ref)) return true;

            if ((string.IsNullOrEmpty(superclass) || superclass == "<nil>") &&
                (interfaces == null || interfaces.Count == 0))
            {
                return false;
            }

            if (!string.IsNullOrEmpty(superclass))
            {
                var success = TraverseTypeHierarchy(superclass);
                if (success) return true;
            }

            if (interfaces != null)
            {
                return interfaces.Select(TraverseTypeHierarchy).Any(success => success);
            }

            return false;
        }


        public override int GetHashCode()
        {
            return HashCode.Combine(base.GetHashCode(), ReturnType, Params, TypeParams);
        }
    }

    public record Union(List<Singular> Types) : Type
    {
        public override string Name()
        {
            return string.Join(" | ", Types.Select(t => t.ToString()));
        }

        public override string ToString()
        {
            return Name();
        }
    }

    public record Parameter(Token ParamName) : Singular(PrimitiveType.Generic, ParamName.Lexeme)
    {
        public override string Name()
        {
            return ParamName.Lexeme;
        }

        public override string ToString()
        {
            return $"type parameter '{ParamName.Lexeme}'";
        }
    }
    
    public static bool Contains(Type type, Type inType, Sunlite? sunlite = null)
    {
        sl = sunlite;
        if (type == Unknown) return false;
        if (inType is Union inU)
        {
            if (type is Union u)
            {
                return inU.Types.SequenceEqual(u.Types) || inU.Types.Contains(Any);
            }

            return inU.Types.Contains(type) || inU.Types.Contains(Any);
        }
        else
        {
            if (type is Union u)
            {
                return u.Types.Contains(type) || inType == Any;
            }

            return inType.Equals(type) || inType == Any;
        }
    }

    public static Type FromValue(object? value, Sunlite sunlite)
    {
        return value switch
        {
            Type type => type,
            Param param => param.Type,
            string => String,
            double => Number,
            bool => Boolean,
            SLNil => Nil,
            SLClosure closure => OfFunction(closure.Function.Name, closure.Function.ReturnType, closure.Function.Params),
            SLBoundMethod boundMethod => OfFunction(boundMethod.Method.Function.Name, boundMethod.Method.Function.ReturnType, boundMethod.Method.Function.Params),
            SLUpvalue upvalue => FromValue(upvalue.ClosedValue, sunlite),
            SLNativeFunction nativeFunc => OfFunction(nativeFunc.Name, nativeFunc.ReturnType, []),
            SLFunction function => OfFunction(function.Name, function.ReturnType, function.Params),
            SLClass clazz => OfClass(clazz.Name, []),
            SLClassInstance instance => OfObject(instance.Clazz.Name),
            SLArray => Array,
            SLTable => Table,
            SLType slType => slType.Value,
            _ => Unknown
        };
    }

    public static Type Of(List<TypeToken> tokens, Sunlite sunlite, bool topmost = true)
    {
        switch (tokens.Count)
        {
            case 0:
                throw new ArgumentException("No types provided.");
            case > 1 when topmost:
                throw new ArgumentException("There can be only one topmost type.");
        }

        if (topmost)
        {
            (Dictionary<Token, List<TypeToken>> typeTokens, List<TypeToken> typeParameters) = tokens.First();
            if (typeTokens.Count == 1)
            {
                KeyValuePair<Token, List<TypeToken>> singleToken = typeTokens.First();
                switch (singleToken.Key.Type)
                {
                    case TokenType.TypeFunction:
                        var returnType = typeParameters.Count == 0 ? Nil : Of([typeParameters.Last()], sunlite, false);
                        List<TypeToken> paramTokens = typeParameters.Take(typeParameters.Count - 1).ToList();
                        List<Param> parameters = paramTokens.Select(t => new Param(Token.Identifier("",""), Of([t], sunlite, false))).ToList();
                        return OfFunction("", returnType, parameters);
                    case TokenType.TypeArray:
                        var elementType = typeParameters.Count == 0
                            ? NullableAny
                            : Of(typeParameters, sunlite, false);
                        return OfArray(elementType);
                    case TokenType.TypeTable:
                        var keyType = typeParameters.Count == 0
                            ? NullableAny
                            : Of([typeParameters[0]], sunlite, false);
                        var valueType = typeParameters.Count == 0
                            ? NullableAny
                            : Of([typeParameters[1]], sunlite, false);
                        return OfTable(keyType, valueType);
                    case TokenType.TypeClass:
                        if(typeParameters.Count == 0)
                        {
                            return OfClass("",[]);
                        }
                        var className =
                            typeParameters.First().Tokens.First().Key.Lexeme;
                        return OfClass(className,[]);
                    case TokenType.Identifier:
                        if (typeParameters.Count > 0)
                        {
                            List<Type> typeParamTypes = typeParameters.Select(tp => Of([tp], sunlite, false)).ToList();
                            var name = singleToken.Key.Lexeme;
                            List<string>? typeParams = sunlite.Collector?.TypeHierarchy.TryGetValue(name, out var tuple) == true ? tuple.Item3 : null;
                            return OfGenericObject(
                                singleToken.Key.Lexeme,
                                typeParamTypes.Select(
                                    (t, i) => 
                                        new Param(
                                            Token.Identifier(
                                                typeParams != null && i < typeParams.Count ? typeParams[i] : "?", 
                                                "<unknown>"), 
                                            t)).ToList());
                        }
                        return OfObject(singleToken.Key.Lexeme);
                    case TokenType.TypeGeneric:
                        return new Parameter(typeParameters.First().Tokens.First().Key);
                    default:
                        return new Singular(PrimitiveTypeExtensions.Get(singleToken.Key));
                }
            }

            List<Singular> types = typeTokens.Values.Select(t => (Of(t, sunlite, false) as Singular)!).ToList();
            return new Union(types);
        }

        if (tokens.Count == 1)
        {
            (Dictionary<Token, List<TypeToken>> typeTokens, List<TypeToken> typeParameters) = tokens.First();
            if (typeTokens.Count == 1)
            {
                KeyValuePair<Token, List<TypeToken>> singleToken = typeTokens.First();
                switch (singleToken.Key.Type)
                {
                    case TokenType.TypeFunction:
                        var returnType = typeParameters.Count == 0 ? Nil : Of([typeParameters.Last()], sunlite, false);
                        List<TypeToken> paramTokens = typeParameters.Take(typeParameters.Count - 1).ToList();
                        List<Param> parameters = paramTokens.Select(t => new Param(Token.Identifier("",""), Of([t], sunlite, false))).ToList();
                        return OfFunction("", returnType, parameters);
                    case TokenType.TypeArray:
                        var elementType = typeParameters.Count == 0
                            ? NullableAny
                            : Of(typeParameters, sunlite, false);
                        return OfArray(elementType);
                    case TokenType.TypeTable:
                        var keyType = typeParameters.Count == 0
                            ? NullableAny
                            : Of([typeParameters[0]], sunlite, false);
                        var valueType = typeParameters.Count == 0
                            ? NullableAny
                            : Of([typeParameters[1]], sunlite, false);
                        return OfTable(keyType, valueType);
                    case TokenType.TypeClass:
                        if (typeParameters.Count == 0)
                        {
                            return OfClass("", []);
                        }

                        var className =
                            typeParameters.First().Tokens.First().Key.Lexeme;
                        return OfClass(className, []);
                    case TokenType.Identifier:
                        if (typeParameters.Count > 0)
                        {
                            List<Type> typeParamTypes = typeParameters.Select(tp => Of([tp], sunlite, false)).ToList();
                            var name = singleToken.Key.Lexeme;
                            List<string>? typeParams = sunlite.Collector?.TypeHierarchy.TryGetValue(name, out var tuple) == true ? tuple.Item3 : null;
                            return OfGenericObject(
                                singleToken.Key.Lexeme,
                                typeParamTypes.Select(
                                    (t, i) => 
                                        new Param(
                                            Token.Identifier(
                                                typeParams != null && i < typeParams.Count ? typeParams[i] : "?", 
                                                "<unknown>"), 
                                            t)).ToList());
                        }
                        return OfObject(singleToken.Key.Lexeme);
                    case TokenType.TypeGeneric:
                        return new Parameter(typeParameters.First().Tokens.First().Key);
                    default:
                        return new Singular(PrimitiveTypeExtensions.Get(singleToken.Key));
                }
            }

            List<Singular> types = typeTokens.Select(t => (Of(t.Value, sunlite, false) as Singular)!).ToList();
            return new Union(types);
        }
        else
        {
            List<Singular> types = tokens.Select(t => (Of([t], sunlite, false) as Singular)!).ToList();
            return new Union(types);
        }
    }
    
    public static Reference OfClass(string name, List<Param> parameters)
    {
        return new Reference(PrimitiveType.Class, name, OfObject(name), parameters, []);
    }

    public static Reference OfObject(string name)
    {
        var reference = new Reference(PrimitiveType.Object, name, Object, [], []);
        return new Reference(PrimitiveType.Object, name, reference, [], []);
    }
    
    public static Reference OfGenericObject(string name, List<Param> typeParams)
    {
        var reference = new Reference(PrimitiveType.Object, name, Object, [], typeParams);
        return new Reference(PrimitiveType.Object, name, reference, [], typeParams);
    }

    public static Reference OfFunction(string name, Type returnType, List<Param> parameters)
    {
        return new Reference(PrimitiveType.Function, name, returnType, parameters, []);
    }

    public static Reference OfGenericFunction(string name, Type returnType, List<Param> parameters, List<Param> typeParams)
    {
        return new Reference(PrimitiveType.Function, name, returnType, parameters, typeParams);
    }

    public static Reference OfArray(Type elementType)
    {
        return new Reference(PrimitiveType.Array, "<array>", elementType, [], []);
    }

    public static Reference OfTable(Type keyType, Type valueType)
    {
        return new Reference(PrimitiveType.Table, "<table>", valueType, [], [new Param(Token.Identifier("<key>",""), keyType)]);
    }

    public static readonly Singular Unknown = new(PrimitiveType.Unknown);
    public static readonly Singular Nil = new(PrimitiveType.Nil);
    public static readonly Singular Any = new(PrimitiveType.Any);
    public static readonly Union NullableAny = new([Any, Nil]);
    public static readonly Singular Number = new(PrimitiveType.Number);
    public static readonly Singular String = new(PrimitiveType.String);
    public static readonly Singular Boolean = new(PrimitiveType.Boolean);
    public static readonly Singular Function = new(PrimitiveType.Function);
    public static readonly Singular Class = new(PrimitiveType.Class);
    public static readonly Singular Object = new(PrimitiveType.Object);
    public static readonly Singular Array = new(PrimitiveType.Array);
    public static readonly Singular Table = new(PrimitiveType.Table);

    public static Sunlite? sl = null;
}